project-chip/connectedhomeip
# Fixing few issues with BindingManager and issues with how clients used BindingManager

#### Problem
* Fixes #21606

#### Change overview
* Fixed
  * Issue where BindingManager::EstablishConnection only called once
  * Prevented used after free
  * Prevented buffer overrun

#### Testing
* CI passes
* Ran the following sequence:
  * Build all clusters app with command `scripts/examples/gn_build_example.sh examples/all-clusters-app/linux out/debug/standalone chip_config_network_layer_ble=false chip_build_libshell=true`
  * `rm -f rm -f kvs1 kvs2 kvs3`
  * `Terminal 1` run `./out/debug/standalone/chip-all-clusters-app --secured-device-port 72564 --KVS kvs1` (port was picked by mashing keys)
  * `Terminal 2` run `./out/debug/standalone/chip-tool pairing onnetwork 1 20202021`
  * `Terminal 3` run `./out/debug/standalone/chip-all-clusters-app --secured-device-port 72574 --KVS kvs2`
  * `Terminal 2` run `./out/debug/standalone/chip-tool pairing onnetwork 2 20202021`
  * `Terminal 4` run `./out/debug/standalone/chip-all-clusters-app --KVS kvs3`
  * `Terminal 2` run `./out/debug/standalone/chip-tool pairing onnetwork 3 20202021`
  * `Terminal 2` run `./out/debug/standalone/chip-tool binding write binding '[{"fabricIndex": 1, "node": 2, "endpoint": 1, "cluster": 6}, {"fabricIndex": 1, "node": 3, "endpoint": 1, "cluster": 6}]' 1 1`
  * Confirm on Terminal 3 and and Terminal 4 that come connection was established to node 1
  * Stop running the all clusters application on Terminal 3 and and Terminal 4
  * `Terminal 1` run `switch on`
    * Wait until you see that the messages were dropped after the max number of re-tries and that the CASESession to node 2 and 3 are now in a defunct state
  * `Terminal 3` run `./out/debug/standalone/chip-all-clusters-app --secured-device-port 72574 --KVS kvs2`
  * `Terminal 4` run `./out/debug/standalone/chip-all-clusters-app --KVS kvs3`
  * `Terminal 1` run `switch on`
  * Confirm on Terminal 3 and and Terminal 4 that come connection was established to node 1 and that it got the notification
